---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```

# Overview

This package provides utility functions to query and explore the expression
profiling data made available through the [ARCHS4 project][archs4web], which
is described in the following publication:
[Massive mining of publicly available RNA-seq data from human and mouse][archs4pub]. 

# Installation

The user will first have to download the four gene and transcript level hdf5
data files made available on the [ARCHS4 downloads][archs4dl] page for the
mouse and human data. These data files need to all be collated into the
directory specified by `getOption("archs4.datadir")` (`~/.archs4data`, by
default).

**NOTE:** If you are developing this package and builidng the documentation, the
build happens in a vanilla R workspace, which won't set your R's `options`
if they are in your `~/.Rprofile`. In this case, symlink your archs4 data
directory such that `~/.archs4data` points to the directory you picked on
your machine.

[archs4web]: https://amp.pharm.mssm.edu/archs4/
[archs4dl]: https://amp.pharm.mssm.edu/archs4/download.html
[archs4pub]: https://www.nature.com/articles/s41467-018-03751-6

# Usage

Please note again that successful usage of this package requires a valid entry
for `getOption("archs4.datadir")`, otherwise many functions will require you to
explicitly pass in a `datadir` parameter.

```{r, message=FALSE, warning=FALSE}
library(archs4)

a4 <- Archs4Repository()
ids <- c('GSE89189', 'GSE29943', "GSM1095128", "GSM1095129", "GSM1095130")
sample.info <- sample_info(a4, ids)
head(sample.info)
```

You can use the `as.DGEList` function to materialize a `edgeR::DGEList` from a
vector of series or sample-level identifiers, as long as they all reference the
same species.

The most often use-case will likely be to create a `DGEList` for a given study.
For instance, the GEO series identifier [`"GSE89189"`][blurtongeo] refers to the
expression data generated to support the
[Abud et al. iPSC-Derived Human Microglia-like Cells ...][blurtonpub] paper.

You can get a `DGEList` of gene-level counts for that study using the command
below. This will create a `DGEList` of 27,024 genes across 37 samples in about
1.5 seconds:

```{r, eval = FALSE}
yg <- as.DGEList(a4, "GSE89189", feature_type = "gene")
```

The following command retrieives the 178,135 transcript level counts for this
experiment in about 1.5 seconds, as well:

```{r, eval = FALSE}
yt <- as.DGEList(a4, "GSE89189", feature_type = "transcript")
```


[blurtonpub]: https://www.ncbi.nlm.nih.gov/pubmed/28426964
[blurtongeo]: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE89189